### PR TITLE
chore: regression test refactor Feb

### PR DESCRIPTION
**Pull request by timstackblock**
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/stacks-wallet-web/actions/runs/1789998377).

Hey team this PR includes 33 new test for bugs we have encountered recently, recommendations by Kyran for functionality that we should test and the addition of test from the automation test documentation in Notion.

MAGIC RECOVERY CODE TEST:
Test users can log in with a magic recover code

LOCKED WALLET TEST:
Test locked wallets display unlocks screen instead of white screen “Stacksart bug”

TRANSACTION SIGNING:
Test transaction signing for token send

ACTIVITY FEED BUG:
Test activity page displays pending transaction

UPDATE TEST TO SKIP METRIC:
Test now select deny button for all automation test to exclude skewing metrics

TRANSACTION FEES TESTING:
For STX Token:
Verify transaction fees work when fees is low
Verify transaction fees work when fees is standard
Verify transaction fees work when fees is high
Verify transaction fees work when fees is custom
validates that custom fee has more than 6 decimal places
validates that custom fee must be positive
validates that the memo must be less than 34-bytes
For STELLA Token:
Verify transaction fees work when fees is low
Verify transaction fees work when fees is standard
Verify transaction fees work when fees is high
Verify transaction fees work when fees is custom
validates that custom fee has more than 6 decimal places
validates that custom fee must be positive
validates that the amount must be more than zero
validates that the memo must be less than 34-bytes
validates that the amount field can have only 9 decimals

ERROR MESSAGING:
Validates that the amount field can only be number
Validates that the address is for the incorrect Stacks network
Validates that amount is greater than the available balance

BUY BUTTON:
Transak test:
Test buy button redirect load and take user to 3rd part site. Clicking button should redirect to transak URL

CONTRACT CALL TRANSACTION FEES:
Verify contract call works when fees is low
Verify contract call works when fees is standard
Verify contract call workswhen fees is high
Verify contract call works when fees is custom
Error messages:
validates that custom fee has more than 6 decimal places
validates that custom fee must be positive
validates that the address used is invalid
validates that the address is same as sender
validates that the amount must be number

cc/ @kyranjamie @fbwoolf @andresgalante @Eshwari007 @beguene @He1DAr

### CODE REVIEW

**Comment by timstackblock** [comment]
Hey guys lets push these refactored test up to get the test pushed in 2 chunks. I have removed the 2 files below and will bring them back once they are refactored to match the changes on the UI that came in recently.

tests/integration/transfer-fees/transfer-fees.spec.ts
tests/integration/contract-call/contract-call.spec.ts

File: src/app/common/error-messages.ts [hunk 1]
```diff
@@ -4,10 +4,12 @@ export enum SendFormErrorMessages {
...
   AmountRequired = 'You must specify an amount',
   MustNotBeZero = 'Must be more than zero',
+  MustBeNumber = 'Amount of STX must be a number',
   DoesNotSupportDecimals = 'This token does not support decimal places',
   InsufficientBalance = 'Insufficient balance. Available:',
   MustSelectAsset = 'You must select a valid token to transfer',
   TooMuchPrecision = '{token} can only have {decimals} decimals',
   MemoExceedsLimit = 'Memo must be less than 34-bytes',
   AdjustedFeeExceedsBalance = 'The fee added now exceeds your current STX balance.',
+  CastToNumber = 'amount must be a `number` type, but the final value was: `NaN`',
```
Comment (fbwoolf) [inline src/app/common/error-messages.ts]: It looks like these error messages are typically started with a capitalized word, is there a reason this one would be different?

File: src/app/components/event-card.tsx [hunk 2]
```diff
@@ -25,7 +26,7 @@ export function EventCard(props: EventCardProps): JSX.Element {
       <Stack p="base-loose" spacing="base-loose">
         <SpaceBetween position="relative">
           <Text fontSize={2} fontWeight={500}>
-            {title}
+            <span data-testid={SendFormSelectors.TransferMessage}>{title}</span>
```
Comment (fbwoolf) [inline src/app/components/event-card.tsx]: This should be moved to the Text component. It is already a span tag.

File: src/app/components/tx-item.tsx [hunk 3]
```diff
@@ -40,7 +41,11 @@ const Status = ({ transaction, ...rest }: { transaction: Tx } & BoxProps) => {
   return isFailed || isPending ? (
     <Box {...rest}>
       {isPending && (
-        <Caption variant="c2" color={color('feedback-alert')}>
+        <Caption
+          variant="c2"
+          color={color('feedback-alert')}
+          data-testid={SendFormSelectors.PendingStatus}
```
Comment (fbwoolf) [inline src/app/components/tx-item.tsx:46]: As a best practice, I try to alphabetize props.

File: src/app/pages/send-tokens/components/asset-search/asset-search.tsx [hunk 4]
```diff
@@ -148,6 +148,7 @@ const AssetSearchField = memo(
             onFocus={() => {
               openMenu();
             }}
+            data-testid={`asset-select`}
```
Comment (fbwoolf) [inline src/app/pages/send-tokens/components/asset-search/asset-search.tsx:151]: Is there a reason sometimes you add selector constants but other times you don't? It seems like we should always try to add selector constants for new tests?

File: test-app/src/components/debugger.tsx [hunk 5]
```diff
@@ -328,7 +329,9 @@ export const Debugger = () => {
       {txId && (
         <Text textStyle="body.large" display="block" my={'base'}>
           <Text color="green" fontSize={1}>
-            Successfully broadcasted &quot;{txType}&quot;
+            <span data-testId={WalletPageSelectors.StatusMessage}>
```
Comment (fbwoolf) [inline test-app/src/components/debugger.tsx]: This should be moved to the Text component.

File: tests/integration/onboarding.selectors.ts [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
   AnalyticsAllowBtn = 'analytics-allow-btn',
+  AnalyticsDenyBtn = 'analytics-deny-btn',
   BackUpSecretKeyBtn = 'back-up-secret-key-btn',
   NewPasswordInput = 'set-or-enter-password-input',
   ConfirmPasswordInput = 'confirm-password-input',
   SetPasswordBtn = 'set-password-btn',
   SignInLink = 'sign-in-link',
   SignInBtn = 'sign-in-btn',
-  SignUpBtn = 'sign-up-btn',
+  SignUpBtn = 'sign-up',
```
Comment (fbwoolf) [inline tests/integration/onboarding.selectors.ts]: Why is this being removed?

File: tests/page-objects/send-form.selectors.ts [hunk 7]
```diff
@@ -7,12 +7,32 @@ export enum SendFormSelectors {
...
-  FeeEstimateItem = 'fee-estimate-item',
+  MemoField = 'memo-field',
+  MemoFieldErrorLabel = 'memo-field-error-label',
+
+  FeeEstimateItem = 'Standard-Fees',
   FeeEstimateSelect = 'fee-estimate-select',
+  LowFeesOption = 'Low-Fees',
+  HighFeesOption = 'High-Fees',
+  StandardFeesOption = 'Standard-Fees',
+  CustomFeesOption = 'Custom-Fees',
```
Comment (fbwoolf) [inline tests/page-objects/send-form.selectors.ts]: I think all these should be singular? `LowFeeOption`, etc. ...`Standard-Fee`, `Low-Fee`, etc.

File: tests/page-objects/send-form.page.ts [hunk 8]
```diff
@@ -95,4 +120,77 @@ export class SendPage {
     await this.page.waitForSelector('[data-asset="stx"]');
     await this.page.click('[data-asset="stx"]');
   }
+
+  async clickSTXTokenOption() {
```
Comment (fbwoolf) [inline tests/page-objects/send-form.page.ts:123]: I search for all these but I don't see any tests using them? Were they lost rebasing?

**Review by fbwoolf — COMMENTED** [review]
~~I think maybe you lost some tests in your rebase, or am I missing something? I only see 5 new tests being added here?~~

**Comment by fbwoolf** [comment]
> Hey guys lets push these refactored test up to get the test pushed in 2 chunks. I have removed the 2 files below and will bring them back once they are refactored to match the changes on the UI that came in recently.
> 

I see this now, nvm my comment about the missing tests!

**Comment by fbwoolf** [comment]
I think it would be best to track tests you are working on in an issue so the team is aware of what you are working on. Is there an issue that can be linked to this PR?

File: tests/page-objects/send-form.selectors.ts (hunk 7 above)
Comment (kyranjamie) [reply tests/page-objects/send-form.selectors.ts]: Also, some of these keys duplicate others in the list, "Standard-Fees" exists twice. Could you remove this duplicate?

**Comment by kyranjamie** [comment]
Thanks for this @timstackblock, and @He1DAr

I rebased the commits into 1. If you can just address Fara's comments, then this is good to merge.

File: src/app/components/event-card.tsx (hunk 2 above)
Comment (timstackblock) [reply src/app/components/event-card.tsx]: The text components make the test fail intermittently, the code above makes the test less flaky.

File: test-app/src/components/debugger.tsx (hunk 5 above)
Comment (timstackblock) [reply test-app/src/components/debugger.tsx]: The text components make the test fail intermittently, the code above makes the test less flaky.

File: src/app/components/event-card.tsx (hunk 2 above)
Comment (fbwoolf) [reply src/app/components/event-card.tsx]: I understand the problem, but I don't agree with this solution ...especially if it is just `less flaky`. We can't add an extra span tag everywhere we use a Text component with a test. Why are we using the Text component here then? What is its purpose?

Comment (kyranjamie) [reply src/app/components/event-card.tsx]: I tested this earlier to make sure there isn't some bug with `<Text/>` not forwarding the `data-testid` prop. It works as expected.

Let's try and isolate what case it is exactly where the tests are flakey. Feel free to ping us when you see this happen @timstackblock, but Fara's concerns are legitimate. `<Text />` renders a `<span>` tag, so here, we're creating unnecessary DOM `<span><span>text</span></span>`.

File: test-app/src/components/debugger.tsx (hunk 5 above)
Comment (kyranjamie) [reply test-app/src/components/debugger.tsx]: @timstackblock Just spotted a capital `I` here.

File: tests/integration/onboarding.selectors.ts (hunk 6 above)
Comment (timstackblock) [reply tests/integration/onboarding.selectors.ts]: This is for the sign up button in the vercel app this matches the element there

File: src/app/components/event-card.tsx (hunk 2 above)
Comment (timstackblock) [reply src/app/components/event-card.tsx]: Yeah I cannot grab the text above it doesn't work. You are welcome to refactor the test after they are merged but I cannot use the text line above with the automated test.

Comment (kyranjamie) [reply src/app/components/event-card.tsx]: @timstackblock take a look at https://github.com/hirosystems/stacks-wallet-web/pull/2207

I've investigated this to see why that happens, and it seems that it works fine.

File: src/app/pages/send-tokens/components/asset-search/asset-search.tsx (hunk 4 above)
Comment (timstackblock) [reply src/app/pages/send-tokens/components/asset-search/asset-search.tsx:151]: Yes we can always add these that is no problem

File: tests/integration/onboarding.selectors.ts (hunk 6 above)
Comment (fbwoolf) [reply tests/integration/onboarding.selectors.ts]: This is used in the `demo.page.ts`, so the selector should just be used there and kept as `sign-up-btn`. The point of using constants is to make sure the string is accurate in all places. There is no point to use a constant in one place and then type out the string again in another.

File: src/app/components/event-card.tsx (hunk 2 above)
Comment (fbwoolf) [reply src/app/components/event-card.tsx]: Thanks for doing this @kyranjamie 👍

File: src/app/pages/send-tokens/components/asset-search/asset-search.tsx (hunk 4 above)
Comment (fbwoolf) [reply src/app/pages/send-tokens/components/asset-search/asset-search.tsx:151]: We should add them before new tests get merged.

File: src/app/components/event-card.tsx (hunk 2 above)
Comment (timstackblock) [reply src/app/components/event-card.tsx]: Thanks @kyranjamie this was helpful

File: src/app/pages/send-tokens/components/asset-search/asset-search.tsx (hunk 4 above)
Comment (timstackblock) [reply src/app/pages/send-tokens/components/asset-search/asset-search.tsx:151]: Ok thanks please add your code change suggestion.

File: test-app/src/components/debugger.tsx (hunk 5 above)
Comment (timstackblock) [reply test-app/src/components/debugger.tsx]: Nice catch I can fix that.

File: src/app/components/tx-item.tsx (hunk 3 above)
Comment (timstackblock) [reply src/app/components/tx-item.tsx:46]: `        <Caption color={color('feedback-alert')} data-testid={SendFormSelectors.PendingStatus} variant="c2">`

Refactoring to this code change please let me know your thoughts

**Comment by timstackblock** [comment]
@fbwoolf @kyranjamie  I just pushed up changes to address the issues above and we are all green feel free to make any edits you see fit on this PR and approve when ready.

**Comment by fbwoolf** [comment]
> @fbwoolf @kyranjamie I just pushed up changes to address the issues above and we are all green feel free to make any edits you see fit on this PR and approve when ready.

@timstackblock thx for making the changes. I do see one other place where there is a `data-testId` with a capital `I`. I can make that change for you and rebase this into 1 commit. If I don't get to it tonight, I will come back to it tomorrow morning. I don't think we want to merge this until after we release?

**Review by fbwoolf — APPROVED** [review]
Made a few changes and rebased into 1 commit. This should be ready now if tests pass.

**Comment by fbwoolf** [comment]
~~@timstackblock the failing test is for `standardFeeSelect` which doesn't seem to exist on the component anywhere? It is just found in a search in the test itself.  Can you fix this or need me to?~~

**Comment by fbwoolf** [comment]
I see the problem, I'll fix it.

**Pull request merged**